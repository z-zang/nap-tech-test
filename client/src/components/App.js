import React, {useState, useEffect} from 'react';

function App() {
  const [products, setProducts] = useState([]);

  async function fetchProducts() {
    var res = await fetch('/api/products')
    var response = await res.json()
    console.log("response", response, typeof response)
    setProducts(response.data)
  }

  useEffect(() => {
    fetchProducts()
  }, [])

  return (
    <div className="App">
      {products.map(el => 
        <div key={el.id}>
          {el.id}
          {el.name}
          {el.price}
          <img src={el.image.outfit} alt={el.name}/>
        </div>
      )}
    </div>
  );
}

export default App;
