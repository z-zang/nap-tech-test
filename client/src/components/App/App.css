body{
  font-family: 'Gill Sans', 'Gill Sans MT', Calibri, 'Trebuchet MS', sans-serif;
  font-size: 1em;
  padding: 10px;
  padding-top: 0px;
  margin: 0px;
}

/* buttons */
.pagination-button, .back-button{
  background-color: black;
  color: white;
  border-radius: 0%;
  font-size: 1em;
  padding: 5px 8px;
}

.pagination-button:disabled, .back-button:disabled{
  background-color: #606060;
  color: whitesmoke;
  border-radius: 0%;
  font-size: 1em;
  padding: 5px 8px;
}

/* so the react router doesn't turn it into hyperlinks */
.router-link{
  text-decoration: none;
  color: black;
}

/* MEDIA QUERIES */

/* tablet */
@media only screen and (min-width: 568px) and (max-width: 834px) {
  body{
    font-size: 0.9em;
  }
  #products-container {
    padding-top: 60px;
  }
  .prodImg{
    width: 160px;
  }
  /* PDP */
  .product-detail{
    grid-template-columns: auto;
  }
}

/* mobile */
@media only screen and (max-width: 568px) {
  body{
    font-size: 0.8em;
  }
  #products-container {
    padding-top: 80px;
  }
  .prodImg{
    width: 130px;
  }
  .product-detail{
    grid-template-columns: auto;
  }
}

/* desktop */
@media only screen and (min-width: 835px) {
  .image-large{
    margin-top: 30px;
    grid-row-start: 1;
    grid-row-end: 11;
  }  
}