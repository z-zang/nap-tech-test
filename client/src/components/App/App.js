import React, {useState, useEffect} from 'react';
import { Switch, Route} from 'react-router-dom'
import Navbar from '../Navbar/Navbar';
import ProductList from '../ProductList/ProductList';
import ProductDetail from '../ProductDetail/ProductDetail';
import Error from '../Error/Error'
import './App.css';

function App() {
  const [products, setProducts] = useState([])
  const [totalProducts, setTotalProducts] = useState('')
  const [offset, setOffset] = useState(0)
  const [priceSort, setPriceSort] = useState('')
  const [selDesigner, setSelDesigner] = useState('')

  // fetch products
  const fetchProducts = async () => {
    let designer = selDesigner === '' ? '' : `&designer=${selDesigner}` 
    let offsetVar = `&offset=${offset}`
    let sort = priceSort === '' ? '' : `&order=${priceSort}` 
    let url = '/api/products?' + designer + sort + offsetVar 
    console.log('fetchProducts requestURL:', url)
    try {
      let res = await fetch(url)
      let response = await res.json()
      setProducts(response.data)
      setTotalProducts(response.total)
    } catch(error) {
      console.log('error', error)
    }
  }

  useEffect(() => {
    fetchProducts()
  }, [offset, priceSort, selDesigner])

  return (
      <React.Fragment>
          <Switch>
            <Route exact path="/" render={(x) => 
              <Navbar {...x} 
                offset={offset} 
                setOffset={setOffset} 
                priceSort={priceSort}
                setPriceSort={setPriceSort}
                selDesigner={selDesigner}
                setSelDesigner={setSelDesigner}
                totalProducts={totalProducts} />}
              />
          </Switch>

          <Switch>
              <Route exact path="/" render={(x) => <ProductList {...x} products={products} totalProducts={totalProducts}/>}/>
              <Route exact path='/product/:pid' component={ProductDetail}/>
              <Route component={Error}/>
          </Switch>
      </React.Fragment>
  );
}

export default App;
