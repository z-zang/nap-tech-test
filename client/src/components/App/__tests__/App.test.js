import React from 'react';
import ReactDOM from 'react-dom';
import App from '../App';
import Adapter from 'enzyme-adapter-react-16';
import { mount, shallow, configure } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

configure({adapter: new Adapter()});

let wrapper;
beforeEach(() => {
  wrapper = mount(
      <App />
    );
});

describe('App', () => {

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.hydrate(wrapper , div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('should match snapshot', async () => {
    expect(wrapper.find(App)).toMatchSnapshot();
  })
})
