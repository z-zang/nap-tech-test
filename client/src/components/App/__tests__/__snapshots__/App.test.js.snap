// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<App>
  <Switch>
    <Route
      computedMatch={
        Object {
          "isExact": true,
          "params": Object {},
          "path": "/",
          "url": "/",
        }
      }
      exact={true}
      location={
        Object {
          "hash": "",
          "key": "01zxbz",
          "pathname": "/",
          "search": "",
          "state": undefined,
        }
      }
      path="/"
      render={[Function]}
    >
      <Navbar
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "canGo": [Function],
            "createHref": [Function],
            "entries": Array [
              Object {
                "hash": "",
                "key": "01zxbz",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
            ],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "index": 0,
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "key": "01zxbz",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
        }
        location={
          Object {
            "hash": "",
            "key": "01zxbz",
            "pathname": "/",
            "search": "",
            "state": undefined,
          }
        }
        match={
          Object {
            "isExact": true,
            "params": Object {},
            "path": "/",
            "url": "/",
          }
        }
        offset={0}
        priceSort=""
        selDesigner=""
        setOffset={[Function]}
        setPriceSort={[Function]}
        setSelDesigner={[Function]}
        totalProducts=""
      >
        <nav>
          <select
            onChange={[Function]}
            value=""
          >
            <option
              value=""
            >
              Sort By
            </option>
            <option
              value="asc"
            >
              Price low to high
            </option>
            <option
              value="desc"
            >
              Price high to low
            </option>
          </select>
          <select
            onChange={[Function]}
            value=""
          >
            <option
              defaultValue={true}
              value=""
            >
              All Designers
            </option>
          </select>
          <span />
          <div>
             result
          </div>
        </nav>
      </Navbar>
    </Route>
  </Switch>
  <Switch>
    <Route
      computedMatch={
        Object {
          "isExact": true,
          "params": Object {},
          "path": "/",
          "url": "/",
        }
      }
      exact={true}
      location={
        Object {
          "hash": "",
          "key": "01zxbz",
          "pathname": "/",
          "search": "",
          "state": undefined,
        }
      }
      path="/"
      render={[Function]}
    >
      <ProductList
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "canGo": [Function],
            "createHref": [Function],
            "entries": Array [
              Object {
                "hash": "",
                "key": "01zxbz",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
            ],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "index": 0,
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "key": "01zxbz",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
        }
        location={
          Object {
            "hash": "",
            "key": "01zxbz",
            "pathname": "/",
            "search": "",
            "state": undefined,
          }
        }
        match={
          Object {
            "isExact": true,
            "params": Object {},
            "path": "/",
            "url": "/",
          }
        }
        products={Array []}
        totalProducts=""
      >
        <main
          id="products-container"
        />
      </ProductList>
    </Route>
  </Switch>
</App>
`;
